File: src/main/java/com/kereyu/hotel/service/RoomService.java
package com.kereyu.hotel.service;

import com.kereyu.hotel.dto.RoomDTO;
import com.kereyu.hotel.exception.ResourceNotFoundException;
import com.kereyu.hotel.model.Room;
import com.kereyu.hotel.repository.RoomRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class RoomService {
    
    @Autowired
    private RoomRepository roomRepository;
    
    public RoomDTO.RoomResponse createRoom(RoomDTO.RoomRequest request) {
        Room room = new Room();
        room.setRoomNumber(request.getRoomNumber());
        room.setRoomType(request.getRoomType());
        room.setPricePerNight(request.getPricePerNight());
        room.setCapacity(request.getCapacity());
        room.setDescription(request.getDescription());
        room.setAmenities(request.getAmenities());
        room.setFloor(request.getFloor());
        room.setImageUrl(request.getImageUrl());
        room.setStatus(Room.RoomStatus.AVAILABLE);
        
        Room savedRoom = roomRepository.save(room);
        return mapToResponse(savedRoom);
    }
    
    public RoomDTO.RoomResponse updateRoom(Long id, RoomDTO.RoomRequest request) {
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Room not found with id: " + id));
        
        room.setRoomNumber(request.getRoomNumber());
        room.setRoomType(request.getRoomType());
        room.setPricePerNight(request.getPricePerNight());
        room.setCapacity(request.getCapacity());
        room.setDescription(request.getDescription());
        room.setAmenities(request.getAmenities());
        room.setFloor(request.getFloor());
        room.setImageUrl(request.getImageUrl());
        
        Room updatedRoom = roomRepository.save(room);
        return mapToResponse(updatedRoom);
    }
    
    public void updateRoomStatus(Long id, Room.RoomStatus status) {
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Room not found with id: " + id));
        room.setStatus(status);
        roomRepository.save(room);
    }
    
    public RoomDTO.RoomResponse getRoomById(Long id) {
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Room not found with id: " + id));
        return mapToResponse(room);
    }
    
    public Page<RoomDTO.RoomResponse> getAllRooms(Pageable pageable) {
        return roomRepository.findAll(pageable).map(this::mapToResponse);
    }
    
    public List<RoomDTO.RoomResponse> getAvailableRooms(LocalDate checkIn, LocalDate checkOut) {
        return roomRepository.findAvailableRooms(checkIn, checkOut).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }
    
    public void deleteRoom(Long id) {
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Room not found with id: " + id));
        roomRepository.delete(room);
    }
    
    private RoomDTO.RoomResponse mapToResponse(Room room) {
        RoomDTO.RoomResponse response = new RoomDTO.RoomResponse();
        response.setId(room.getId());
        response.setRoomNumber(room.getRoomNumber());
        response.setRoomType(room.getRoomType());
        response.setPricePerNight(room.getPricePerNight());
        response.setCapacity(room.getCapacity());
        response.setDescription(room.getDescription());
        response.setAmenities(room.getAmenities());
        response.setFloor(room.getFloor());
        response.setStatus(room.getStatus());
        response.setImageUrl(room.getImageUrl());
        return response;
    }
}

File: frontend/src/app/services/room.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { RoomRequest, RoomResponse } from '../models/room.model';
import { Page } from '../models/reservation.model';

@Injectable({ providedIn: 'root' })
export class RoomService {
  private apiUrl = 'http://localhost:8080/api/rooms';

  constructor(private http: HttpClient) {}

  getAllRooms(page = 0, size = 20): Observable<Page<RoomResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<RoomResponse>>(this.apiUrl, { params });
  }

  getRoomById(id: number): Observable<RoomResponse> {
    return this.http.get<RoomResponse>(`${this.apiUrl}/${id}`);
  }

  getAvailableRooms(checkIn: string, checkOut: string): Observable<RoomResponse[]> {
    const params = new HttpParams().set('checkIn', checkIn).set('checkOut', checkOut);
    return this.http.get<RoomResponse[]>(`${this.apiUrl}/available`, { params });
  }

  createRoom(room: RoomRequest): Observable<RoomResponse> {
    return this.http.post<RoomResponse>(this.apiUrl, room);
  }

  updateRoom(id: number, room: RoomRequest): Observable<RoomResponse> {
    return this.http.put<RoomResponse>(`${this.apiUrl}/${id}`, room);
  }

  updateRoomStatus(id: number, status: string): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/status`, { status });
  }

  deleteRoom(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
